Allow explicit typing on selector using TypeVar with default Any

leonarduschen opened the pull request:
Partially addresses #930

leonarduschen committed:
Use TypeVar with default Any for Selector

leonarduschen committed:
Add tests

ZipFile commented:
Could you rebase with latest `develop` branch? Failing tests should be fixed now (#933).

ZipFile commented on src/dependency_injector/providers.pyi:
```diff
@@ -22,6 +21,8 @@ from typing import (
     overload,
 )
 
+from typing_extensions import TypeVar
```
We need to bump minimal Python version for `typing_extensions` [here](https://github.com/ets-labs/python-dependency-injector/blob/059f78b27ccecabc5f8cf68f181e3253002c8f34/pyproject.toml#L58).

leonarduschen replied:
Right, bumped to v3.13 for TypeVar default

ZipFile commented on src/dependency_injector/providers.pyi:
```diff
@@ -542,17 +544,17 @@ class Container(Provider[T]):
     @property
     def providers(self) -> _Dict[str, Provider]: ...
-    def set_providers(self, **providers: Provider) -> Selector: ...
+    def set_providers(self, **providers: Provider) -> Selector[TT]: ...
```
```suggestion
    def set_providers(self, **providers: Provider) -> Self: ...
```
(`Self` from `typing_extensions`)

leonarduschen replied:
Ah, I think I wanted to use `T_Any` on this one (was going to check everything before converting the PR from "draft" to "ready")

---

Anyway, changed to `Selector[T_Any]` since similar usecases in this module are not using `typing_extensions.Self` either, let me know if you want me to use typing_extensions.Self, the difference will only be when subclassing `Selector`.

leonarduschen replied:
Note:

We'll have a decision to make on `set_providers` later once we make Provider covariant


**Option 1:**

Require `Selector[MyType].set_provider` to only accept the providers assignable to `Provider[MyType]`, and so we don't change the returned Selector type

```python
    def set_providers(self, **providers: Provider[T_Any]) -> Selector[T_Any]: ...
```

**Option 2:**

Allow any kind of providers and use union for the return type

```python
    def set_providers(self, **providers: Provider[TT_Any]) -> Selector[T_Any | TT_Any]: ...
```

---

In this PR we make it `    def set_providers(self, **providers: Provider) -> Selector[T_Any]: ...` (not changing the retuned Selector type) so we're already leaning toward the first option. I personally prefer it this way to, easier to make sense of

ZipFile replied:
> Anyway, changed to `Selector[T_Any]` since similar usecases in this module are not using typing_extensions.Self either, let me know if you want me to use `typing_extensions.Self`, the difference will only be when subclassing `Selector`.

I was thinking about `typing_extensions.Self` from more practical standpoint. This method literally [returns self](https://github.com/ets-labs/python-dependency-injector/blob/5a1aef920372da6a95d14865caab10061aea7104/src/dependency_injector/providers.pyx#L4107-L4110). This should save us some headache in the future.

> Require Selector[MyType].set_provider to only accept the providers assignable to Provider[MyType], and so we don't change the returned Selector type

Looks reasonable enough.

```python
from typing_extensions import Self as _Self  # alias to not collide with Self provider

def set_providers(self, **providers: Provider[T_Any]) -> _Self: ...
```

leonarduschen replied:
Sure thing

leonarduschen committed:
Bump minimal Python version for typing_extensions to v3.13

leonarduschen committed:
Fix accidental return type change on Selector.set_providers

leonarduschen committed:
Change Selector.set_provider to return Selector[T_Any]

leonarduschen committed:
Update tests

leonarduschen commented:
Thanks for the quick review!

I've rebased and addressed your comments @ZipFile.

leonarduschen committed:
Specify Provider type for Selector.__getattr__ and Selector.providers

leonarduschen committed:
Add more tests

leonarduschen commented:
Sorry, just added more changes:
- Specify return type of `Selector.__getattr__` and `Selector.providers`
- Added more tests to make sure untyped Selector is not affected by these changes

leonarduschen committed:
Use typing_extensions.Self on set_selector and set_providers

ZipFile commented:
Thanks!
